eval/src/vespa/eval/instruction/mixed_inner_product_function.cpp

```diff
@@ -0,0 +1,134 @@
+    double result = 0.0;
+    for (size_t i = 0; i < count; ++i) {
+        result += (*lhs++) * (*rhs++);
+    }
+    return result;
+}
+
+template <>
+double my_dot_product<double,double>(const double * &lhs, const double * rhs, size_t count) {
+    double result = cblas_ddot(count, lhs, 1, rhs, 1);
+    lhs += count;
+    return result;
+}
+
+template <>
+double my_dot_product<float,float>(const float * &lhs, const float * rhs, size_t count) {
+    double result = cblas_sdot(count, lhs, 1, rhs, 1);
+    lhs += count;
+    return result;
+}
+
+template <typename MCT, typename VCT, typename OCT>
+void my_mixed_inner_product_op(InterpretedFunction::State &state, uint64_t param_in) {
+    const auto &param = unwrap_param<MixedInnerProductParam>(param_in);
+    const auto &mixed = state.peek(1);
+    const auto &dense = state.peek(0);
+    auto m_cells = mixed.cells().typify<MCT>();
+    auto v_cells = dense.cells().typify<VCT>();
+    const auto &index = mixed.index();
+    size_t num_subspaces = index.size();
+    size_t num_output_cells = num_subspaces * param.out_subspace_size;
+    ArrayRef<OCT> out_cells = state.stash.create_uninitialized_array<OCT>(num_output_cells);
+    const MCT *m_cp = m_cells.begin();
+    OCT *out = out_cells.begin();
+    for (size_t i = 0; i < num_output_cells; ++i) {
+        const VCT *v_cp = v_cells.begin();
+        *out++ = my_dot_product(m_cp, v_cp, param.vector_size);
+    }
+    assert(out == out_cells.end());
+    assert(m_cp == m_cells.end());
+    state.pop_pop_push(state.stash.create<ValueView>(param.res_type, index, TypedCells(out_cells)));
+}
+        
+
+struct SelectMixedInnerProduct {
+    template <typename MCT, typename VCT, typename OCT>
+    static auto invoke() { return my_mixed_inner_product_op<MCT,VCT,OCT>; }
+};
+
+} // namespace <unnamed>
+
+MixedInnerProductFunction::MixedInnerProductFunction(const ValueType &res_type_in,
+                                                     const TensorFunction &lhs_in,
+                                                     const TensorFunction &rhs_in)
+  : tensor_function::Op2(res_type_in, lhs_in, rhs_in)
+{
+}
+
+InterpretedFunction::Instruction
+MixedInnerProductFunction::compile_self(const ValueBuilderFactory &, Stash &stash) const
+{
+    const auto &mix_type = lhs().result_type();
+    const auto &vec_type = rhs().result_type();
+    auto &param = stash.create<MixedInnerProductParam>(result_type(), mix_type, vec_type);
+    using MyTypify = TypifyValue<TypifyCellType>;
+    auto op = typify_invoke<3,MyTypify,SelectMixedInnerProduct>(mix_type.cell_type(),
+                                                                vec_type.cell_type(),
+                                                                result_type().cell_type());
+    return InterpretedFunction::Instruction(op, wrap_param<MixedInnerProductParam>(param));
+}
+
+bool
+MixedInnerProductFunction::compatible_types(const ValueType &res, const ValueType &mixed, const ValueType &vector)
```

Comment:
need to verify that we are not reducing any of the mapped dimensions (also; add test for this)